apps/src/lab2/views/ExtraLinksModal.tsx

```diff
@@ -0,0 +1,174 @@
+import React, {useEffect, useState} from 'react';
+import {ExtraLinksData} from '../types';
+import {Heading3, StrongText} from '@cdo/apps/componentLibrary/typography';
+import AccessibleDialog from '@cdo/apps/templates/AccessibleDialog';
+import moduleStyles from './extra-links.module.scss';
+import Button from '@cdo/apps/templates/Button';
+import {getAuthenticityToken} from '@cdo/apps/util/AuthenticityTokenStore';
+
+// Extra Links modal. This is used to display helpful links for levelbuilders, and should
+// be extended to also include links for project validators as well. It replaces the haml
+// version of extra links, which doesn't work on lab2 after a level change.
+interface ExtraLinksModalProps {
+  linkData: ExtraLinksData;
+  isOpen: boolean;
+  setIsOpen: (isOpen: boolean) => void;
+  levelId: number;
+}
+
+const ExtraLinksModal: React.FunctionComponent<ExtraLinksModalProps> = ({
+  linkData,
+  isOpen,
+  setIsOpen,
+  levelId,
+}) => {
+  const [showCloneField, setShowCloneField] = useState(false);
+  const [showDeleteConfirm, setShowDeleteConfirm] = useState(false);
+  const [clonedLevelName, setClonedLevelName] = useState('');
+  const [cloneError, setCloneError] = useState('');
+  const [deleteError, setDeleteError] = useState('');
+
+  useEffect(() => {
+    if (linkData) {
+      setClonedLevelName(linkData.level_name);
+    }
+  }, [linkData]);
+
+  const onClose = () => {
+    setIsOpen(false);
+    setShowCloneField(false);
+    setShowDeleteConfirm(false);
+    setCloneError('');
+    setDeleteError('');
+  };
+
+  const handleClone = async () => {
+    if (clonedLevelName) {
+      const response = await fetch(
+        `/levels/${levelId}/clone?name=${clonedLevelName}`,
+        {
+          method: 'POST',
+          headers: {
+            contentType: 'application/json;charset=UTF-8',
+            'X-CSRF-Token': await getAuthenticityToken(),
+          },
+        }
+      );
+      if (response.ok) {
```

Comment:
nit: I've typically used both `response.ok` and `try/catch` together, and now I'm wondering why. I think to also catch errors related to network issues or CORS or something and also any errors that don't happen until `response.json()`. I'm not actually sure it's necessary, but something to consider.

Reply:
I'm still getting used to fetch vs ajax, adding a try/catch makes sense!

Reply:
FWIW, `HttpClient` functions should throw even when a response is not `ok`, so just wrapping those in a try/catch should be sufficient!